mnefun/_mnefun.py

```diff
@@ -2330,3 +2332,108 @@ def plot_reconstruction(evoked, origin=(0., 0., 0.04)):
     axs[2, 0].set_ylabel('Reconstruction')
     fig.tight_layout()
     return fig
+
+
+def plot_chpi_snr_raw(raw, win_length, n_harmonics):
+    """Compute and plot cHPI SNR from raw data
+
+    Parameters
+    ----------
+    win_length : int
+        Length of window to use for SNR estimates (seconds). A longer window
+        will naturally include more low frequency power, resulting in lower
+        SNR.
+    n_harmonics : int
+        Number of line frequency harmonics to include in the model.
+
+    Returns
+    -------
+    fig : Instance of matplotlib.figure.Figure
+        cHPI SNR as function of time, residual variance.
+    """
+
+    # plotting parameters
+    legend_fontsize = 10
+    legend_hspace = 25  # % of horizontal axis to reserve for legend
+
+    # get some info from fiff
+    sfreq = raw.info['sfreq']
+    linefreq = raw.info['line_freq']
+    linefreqs = (np.arange(n_harmonics + 1) + 1) * linefreq
+    buflen = int(win_length * sfreq)
+    if buflen <= 0:
+        raise ValueError('Window length should be >0')
+    (cfreqs, _, _, _, _) = _get_hpi_info(raw.info)
+    print('\nNominal cHPI frequencies: %s Hz' % cfreqs)
+    print('Sampling frequency: %s Hz' % sfreq)
+    print('Using line freqs: %s Hz' % linefreqs)
+    print('Using buffers of %s samples = %s seconds\n'
+          % (buflen, buflen/sfreq))
+
+    pick_meg = pick_types(raw.info, meg=True, exclude=[])
+    pick_mag = pick_types(raw.info, meg='mag', exclude=[])
+    pick_grad = pick_types(raw.info, meg='grad', exclude=[])
+    nchan = len(pick_meg)
+
+    # create general linear model for the data
+    t = np.arange(buflen) / float(sfreq)
+    model = np.c_[t, np.ones(t.shape)]  # model slope and DC
+    # add sine and cosine term for each freq
+    for f in np.concatenate([linefreqs, cfreqs]):
+        model = np.c_[model, np.cos(2*np.pi*f*t), np.sin(2*np.pi*f*t)]
```

Comment:
This is memory inefficient and ambiguous about the resulting size (it needs to be mentally calculated rather than being clear from the code), better to preallocate `model = np.empty((len(t), 2 + len(linefreqs) + len(cfreqs)))` and populate it using indices. I know it probably doesn't make a noticable performance difference here, but it's better to get in good habits early :)